src/core/thread/mle.cpp

```diff
@@ -375,6 +375,9 @@ otError Mle::Restore(void)
 
     SuccessOrExit(error = Get<Settings>().ReadNetworkInfo(networkInfo));
 
+    // force re-attach when version mismatch.
+    VerifyOrExit(networkInfo.mVersion == kThreadVersion);
```

Comment:
I would move this check to just before the `switch (networkInfo.Role)` line below.